File: ops/src/infinicore/dtype.cc
#include <infinicore.hpp>

namespace infinicore {

std::string toString(const DataType &dtype) {
    switch (dtype) {
    case DataType::BYTE:
        return "BYTE";
    case DataType::BOOL:
        return "BOOL";
    case DataType::I8:
        return "I8";
    case DataType::I16:
        return "I16";
    case DataType::I32:
        return "I32";
    case DataType::I64:
        return "I64";
    case DataType::U8:
        return "U8";
    case DataType::U16:
        return "U16";
    case DataType::U32:
        return "U32";
    case DataType::U64:
        return "U64";
    case DataType::F8:
        return "F8";
    case DataType::F16:
        return "F16";
    case DataType::F32:
        return "F32";
    case DataType::F64:
        return "F64";
    case DataType::C16:
        return "C16";
    case DataType::C32:
        return "C32";
    case DataType::C64:
        return "C64";
    case DataType::C128:
        return "C128";
    case DataType::BF16:
        return "BF16";
    }

    // TODO: Add error handling.
    return "";
}

size_t dsize(const DataType &dtype) {
    switch (dtype) {
    case DataType::BYTE:
    case DataType::BOOL:
    case DataType::F8:
    case DataType::I8:
    case DataType::U8:
        return 1;
    case DataType::I16:
    case DataType::U16:
    case DataType::F16:
    case DataType::BF16:
    case DataType::C16:
        return 2;
    case DataType::I32:
    case DataType::U32:
    case DataType::F32:
    case DataType::C32:
        return 4;
    case DataType::I64:
    case DataType::U64:
    case DataType::F64:
    case DataType::C64:
        return 8;
    case DataType::C128:
        return 16;
    }

    // TODO: Add error handling.
    return 0;
}

} // namespace infinicore

File: python/infinilm/core/dtype.py
from infinilm.core.lib import _core


class dtype:
    def __init__(self, data_type):
        """An internal method. Please do not use this directly."""
        self._underlying = data_type

    def __repr__(self):
        repr_map = {
            _core.DataType.BYTE: "uint8",
            _core.DataType.BOOL: "bool",
            _core.DataType.I8: "int8",
            _core.DataType.I16: "int16",
            _core.DataType.I32: "int32",
            _core.DataType.I64: "int64",
            _core.DataType.U8: "uint8",
            _core.DataType.U16: "uint16",
            _core.DataType.U32: "uint32",
            _core.DataType.U64: "uint64",
            _core.DataType.F8: "float8",
            _core.DataType.F16: "float16",
            _core.DataType.F32: "float32",
            _core.DataType.F64: "float64",
            _core.DataType.C16: "complex16",
            _core.DataType.C32: "complex32",
            _core.DataType.C64: "complex64",
            _core.DataType.C128: "complex128",
            _core.DataType.BF16: "bfloat16",
        }
        return f"infinicore.{repr_map[self._underlying]}"

    def __eq__(self, other):
        """
        Compare two dtype objects for equality.

        Args:
            other: The object to compare with

        Returns:
            bool: True if both objects are dtype instances with the same underlying data type
        """
        if not isinstance(other, dtype):
            return False
        return self._underlying == other._underlying

    def __hash__(self):
        """
        Return a hash value for the dtype object.

        Returns:
            int: Hash value based on the underlying data type
        """
        return hash(self._underlying)


float32 = dtype(_core.DataType.F32)
float = float32
float64 = dtype(_core.DataType.F64)
double = float64
complex32 = dtype(_core.DataType.C32)
chalf = complex32
complex64 = dtype(_core.DataType.C64)
cfloat = complex64
complex128 = dtype(_core.DataType.C128)
cdouble = complex128
float16 = dtype(_core.DataType.F16)
half = float16
bfloat16 = dtype(_core.DataType.BF16)
uint8 = dtype(_core.DataType.U8)
int8 = dtype(_core.DataType.I8)
int16 = dtype(_core.DataType.I16)
short = int16
int32 = dtype(_core.DataType.I32)
int = int32
int64 = dtype(_core.DataType.I64)
long = int64
bool = dtype(_core.DataType.BOOL)
